ibc/docs/ics24.md

```diff
@@ -27,6 +55,60 @@ Identifiers are bytestrings constrained to specific characters and lengths depen
 
 The `GetTimestamp()` function returns the current unix timestamp of the host machine and is used to calculate timeout periods for packets
 
+## IBC State
+
+As mentioned [above](#persistence) the IBC store **must** be included in the consensus state of the network. As such the IBC store as defined in [ICS-24][ics24] has been implemented as a single IBC state tree.
+
+### IBC State Tree
+
+The IBC state tree is a non-value hashing `SMT` backed by a persistent `KVStore`, this is due to its need for data retrieval as well as proof generation/verification. The root hash of the IBC state tree is included in the `rootTree` which computes the networks state hash for any given block. This allows verifiers to not only verify the inclusion/exclusion of any element in the IBC state tree itself but also that the IBC state tree was used to compute the network's state hash, by utilising the `CommitmentProof` object defined in [ICS-23][ics23].
```

Comment:
If we were to read from a SMT without using the `WithValueHasher(nil)` option in its creation we will **only** get the values hash. I don't see where the current state trees are being used for retrieval but I may be wrong. This PR makes all state trees non-value hashing. In a non value hashing tree when we do `tree.Get([]byte("key"))` we get the actual value, not its hash.

Reply:
So the idea behind the ADR is very clear and makes me think its a good idea. This would be very good to flesh out on a call but the way we interact with the stores will have to change. The following is an idea of how this would work (the emit functions are found in the ICS-24 submodule PR yet to be put up)

```go
type StoreManager struct {
	stores map[string]*ProvableStore
}

func (s *StoreManager) PruneCaches(height uint64) error {
	for _, store := range s.stores {
		if err := store.PruneCache(height); err != nil {
			return err
		}
	}
	return nil
}

// CachedEntry represents a local change made to the IBC store prior to it being
// committed to the state tree. These should be written to disk in the to prevent a
// loss of data and pruned when included in the state tree
// written to disk as follows:
// "{height}/{prefixedKey}" => value
type CachedEntry struct {
	height uint64
	prefixedKey []byte
	value []byte
}

func (c *CachedEntry) prepare() ([]byte, []byte) {
	return []byte(fmt.Sprintf("%d/%s", c.height, string(c.prefixedKey))), value
}

// ProvableStore is a struct that interfaces with the PostgresDB instance
// obtained from Persistence. It is used to Get/Set/Delete the keys in the
// IBC state tree, in doing so it will trigger the creation of 
type ProvableStore struct {
	bus modules.Bus // used to interact with persistence (passed from IBCHost)
	name string
	storeDir string // /var/stores/{name}
	prefix coreTypes.CommitmentPrefix // []byte(name)
	cache []*CachedEntry
}

func (p *ProvableStore) Get(key []byte) ([]byte, error) {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	return persistenceMod.GetIBCEntry(prefixed, int64(currHeight)-1) // returns latest value stored
}

func (p *ProvableStore) GetAndProve(key []byte, membership bool) ([]byte, *ics23.CommitmentProof, error) {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	value, err := persistenceMod.GetIBCEntry(prefixed, int64(currHeight)-1) // returns latest value stored
	if err != nil {
		return nil, nil, err
	}
	// import IBC state tree
	root, nodeStore := p.bus.GetTreeStore()
	lazy := smt.ImportSparseMerkleTree(nodeStore, root, sha256.New())
	proof, err := lazy.Prove(prefixed)
	if err != nil {
		return nil, nil, err
	}
	// convert SMT proof into CommitmentProof
	ics23Proof := new(ics23.CommitmentProof)
	if membership {
		ics23Proof, err = createMembershipProof(proof, prefixed, value)
	} else {
		ics23Proof, err = createNonMembershipProof(proof, prefixed)
	}
	return value, ics23Proof, nil
}

func (p *ProvableStore) Set(key, value []byte) error {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	if err := persistenceMod.SetIBCEntry(prefixed, value); err != nil {
		return err
	}
	return emitUpdateStoreEvent(p.prefix, key, value)
}


func (p *ProvableStore) Delete(key []byte) error {
	prefixed := applyPrefix(p.prefix, key)
	persistenceMod := p.bus.GetPersistenceModule()
	currHeight := p.bus.GetConsensusModule().CurrentHeight()
	if err := persistenceMod.SetIBCEntry(prefixed, nil); err != nil {
		return err
	}
	return emitPruneStoreEvent(p.prefix, key, value)
}

func (p *ProvableStore) CacheEntries() error {
	store, err := kvstore.NewKVStore(p.storeDir)
	if err != nil {
		return err
	}
	for _, entry := range p.cache {
		key, value := entry.prepare()
		if err := store.Set(key, value); err != nil {
			return err
		}
	}
	p.cache := make([]*CachedEntry, 0)
	return store.Stop()
}

func (p *ProvableStore) PruneCache(height uint64) error {
	store, err := kvstore.NewKVStore(p.storeDir)
	if err != nil {
		return err
	}
	keys, _, err := store.GetAll([]byte(fmt.Sprintf("%d", height)), false)
	if err != nil {
		return err
	}
	for _, key := range keys {
		if err := store.Delete(key); err != nil {
			return err
		}
	}
	return store.Stop()
}
```

Reply:
@h5law def let's hop on a call today about this 👍